modules/rabbitmq/rabbitmanager.go

```diff
@@ -279,30 +274,28 @@ func RunSubscriber(cfg *AMQPSubCfg) error {
 
 // RunPublisher creates a new publisher, and continually pushes messages submitted to the Go channel
 // to RabbitMQ.
-func RunPublisher(cfg *AMQPPubCfg) error {
-	if cfg.Control == nil {
-		cfg.Control = utils.NewControl()
-	}
+func RunPublisher(cfg *AMQPPubSubCfg) error {
 	defer func() {
-		close(cfg.Messages)
-		close(cfg.Control.Exit)
+		close(cfg.PubCfg.Messages)
+		cfg.shutdown.Do(func() { // Make sure this is only ever called once.
+			close(cfg.Control.Exit) // If subscriber exits, kill publisher as well.
+		})
 	}()
 
 	ch, err := GetChannel()
 	if err != nil {
-		utils.LogError("Failed to get new channel", err, nil)
-		// panic so we shut down the subscriber too
-		panic(err) // TODO(shapiro): Think of a better way of having publisher and consumer be able to shut each other down
+		// Shut down subscriber if failed here.
+		return fmt.Errorf("RunPublisher: Failed to get new channel: %v", err)
```

Comment:
Nah. this gets logged at a higher level. We probably don't want to pollute the logs by logging twice.